Build_Process/BootProtocols.md

```diff
@@ -0,0 +1,349 @@
+
+```c
+void kernel_start(stivale2_struct* stivale2_data);
+```
+
+This struct points to a list of tags, each containing details about the machine we're booted on. These are called struct tags (bootloader -> kernel) as opposed to the tags we defined before (header tags: kernel -> bootloader). To get info about a specific feature, simply walk the linked list of tags, the next tag's address is available in the `tag->next` field. The end of the list is indicated by a nullptr.
+
+## Finding Bootloader Tags
+Since both multiboot 2 and stivale 2 return their info in linked lists, a brief example of how to traverse these lists is given below. These functions provide a nice abstraction to search the list for a specific tag, rather than manually searching each time.
+
+### Multiboot 2
+Multiboot 2 gives us a pointer to the multiboot info struct, which contains 2x 32-bit fields. These can be safely ignored, as the list is null-terminated (a tag with a type 0, and size of 8). The first tag is at 8 bytes after the start of the mbi. All the structures and defines used here are available in the header provided by the multiboot specification (check the bottom section, in the example kernel).
+
+```c
+//placed in ebx when the kernel is booted
+multiboot_info* mbi;
+
+void* multboot2_find_tag(uint32_t type)
+{
+    multiboot_tag* tag = (multiboot_tag*)(uintptr_t)mbi + 8);
+    while (1)
+    {
+        if (tag->type == 0 && size == 8)
+            return NULL; //we've reached the terminating tag
+        
+        if (tag->type == type)
+            return tag;
+        
+        tag = (multiboot_tag*)((uintptr_t)tag + tag->size);
+    }
+}
+```
+
+The last line of the loop is a little messy, in fact most compilers will let you omit the final cast to the `void*`, and will convert the integer expression directory to a pointer for you (gcc and clang both allow this).
+
+### Stivale 2
+Stivale 2 gives us a pointer to a header at the start of the list, and then each item (including this header) contains a `next` pointer to the next item, and an `id` item with a unique 64-bit identifier for that tag. All the structures and defines are available in the standard `stivale2.h`. We'll know we've reached the end of the list when the `next` pointer is `NULL`.
+
+```c
+//given to the kernel entry function
+stivale2_struct* s2_struct;
+
+//returns null if tag could not be found
+void* stivale2_find_tag(uint64_t id)
+{
+    stivale2_tag* tag = s2_struct->next;
+
+    while (tag != NULL)
+    {
+        if (tag->id == id)
+            return tag;
+        tag = tag->next;
+    }
+
+    return NULL;
+}
+```
+
+The above function can be used with the defines in `stivale2.h`, which follow the format `STIVALE2_STRUCT_TAG_xyz_ID`, where `xyz` represents the feature that is described in the tag. For example, the framebuffer would be `STIVALE2_STRUCT_TAG_FRAMEBUFFER_ID` and the memory map is `STIVALE2_STRUCT_TAG_MEMMAP_ID`. It's a little verbose, but easy to search for.
```

Comment:
maytbe is worth adding a similar sentence on the multiboot paragraph too? mentioning that there is a header available ready to use, and all the types and multiboot struct are declared there?